### PR TITLE
[SYCL][Doc] Add spec for "spirv" to kernel compiler

### PR DESCRIPTION
**Pull request by gmlueck**
Add an extension that defines support for online compiled kernels that
are written in binary SPIR-V.

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler_spirv.asciidoc [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+[%header,cols="1,1"]
+|===
+|SPIR-V type
+|Corresponding SYCL type
+
+|*OpTypeInt*
+|A C++ integral type with the same width.
+
+|*OpTypeFloat* of width 32
+|`float`
```
Comment (bsochack) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler_spirv.asciidoc:187]: Is there any reason why dtypes are mentioned here? As long as it is supported by SPIR-V, it should be fine.
There are other dtypes that are important for us and they are not mentioned i.e. bfloat16 or float16/half

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc [hunk 2]
```diff
@@ -172,22 +172,33 @@ a!
...
 The bundle is associated with the context `ctxt`, and kernels from this bundle
 may only be submitted to a queue that shares the same context.
 The bundle's set of associated devices is the set of devices contained in
 `ctxt`.
 
+_Effects (2):_ Same as (1) except the source code is defined as a vector of
+bytes.
+This overload is provided for the benefit of source languages that are more
+naturally expressed as a series of raw bytes.
+However, both overloads can be used for any supported language.
```
Comment (Pennycook) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc:200]: What would this actually mean if somebody were to pass in a SYCL source program (for example) as a series of raw bytes?  Would it be the raw bytes representing a string?  Would there be any restrictions on the encoding in that case? (e.g., would it support ASCII and UTF-8, or is it up to the implementation to somehow communicate what the bytes must be?)

**Comment by jbrodman** [comment]
Maybe this is opening too big a can of worms - but it's a little odd that we're calling binary SPIRV "source". In OpenCL there is createProgramFromSource AND createProgramFromIL.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc (hunk 2 above)
Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc]: I was waffling on this decision, but I think you're right.  I don't want to specify the character encoding when reading a textual language from a vector of bytes.  Instead, I classify each source language as either text format or binary format and make it the application's responsibility to call the right overload.  See 63c8460db1acb9de3d8cfd7b55eb174cf5f7591c.

**Comment by gmlueck** [comment]
> Maybe this is opening too big a can of worms - but it's a little odd that we're calling binary SPIRV "source". In OpenCL there is createProgramFromSource AND createProgramFromIL.

I'm, standing firm on this one.  I think the important distinction is whether the language is text format or binary format, and we can deal with this nicely by having two overloads of `create_kernel_bundle_from_source`.  Note that OpenCL is C based, so they did not have the option to expose different overloads of `clCreateProgram`.

I also think the distinction of "source" and "IL" could get fuzzy.  For example, what if we want to allow kernels to be defined using the textual form of SPIR-V in the future?  Would we call this "source" (and not "IL") just because we need to run the program through the SPIR-V assembler?  I think it's better to expose just a single function name `create_kernel_bundle_from_source` that has two overloads.

**Review by Pennycook — APPROVED** [review]

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc (hunk 2 above)
Comment (Pennycook) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc]: I think this is a good solution.  If somebody really wants to store an OpenCL/SYCL string in byte form, it's not that hard for them to load it into a `std::string` before compilation.

**Comment by gmlueck** [comment]
@intel/llvm-gatekeepers I think this is ready to merge.

I talked with @jbrodman offline, and he does not feel so strongly about his comment that he opposes this PR.

**Pull request merged**